File: Feladat3/sort/QuickSort.cs
using System;

namespace Feladat3.alg {
    
    public class QuickSort {

        public static void Sort(int[] array)
            => quicksort(array, 0, array.Length - 1);
        

        private static void quicksort(int[] array, int start, int end) {
            if (end > start) {
                int pivot = findGoodPivot(array, start, end);
                int newPivot = partition(array, start, end, pivot);
                quicksort(array, start, newPivot - 1);
                quicksort(array, newPivot + 1, end);
            }
        }
        
        private static int findGoodPivot(int[] array, int start, int end) { 
            int aIdx = start;
            int a = array[aIdx];
            int bIdx = (start + end) / 2;
            int b = array[bIdx];
            int cIdx = end;
            int c = array[cIdx];
            
            if (a > b) { swap(ref a, ref b); swap(ref aIdx, ref bIdx); }
            if (b > c) { swap(ref b, ref c); swap(ref bIdx, ref cIdx); }
            if (a > b) { swap(ref a, ref b); swap(ref aIdx, ref bIdx); }

            return bIdx;

        }

        
        private static int partition(int[] array, int start, int end, int pivot) {
            int pivotVal = array[pivot];
            swap(array, pivot, end);
            int storeIdx = start;
            for (int i = start; i < end; i++) {
                if (array[i] < pivotVal) { // array[i].CompareTo(pivotVal) < 0
                    swap(array, i, storeIdx);
                    storeIdx++;
                }
            }
            swap(array, storeIdx, end);
            return storeIdx;
        }
        

        private static void swap(int[] array, int idx1, int idx2) {
            int temp = array[idx1];
            array[idx1] = array[idx2];
            array[idx2] = temp;
        }

        private static void swap(ref int a, ref int b) {
            int temp = a;
            a = b;
            b = temp;
        }
        
    }
}

File: Feladat4/src/sort/QuickSort.kt
package sort


fun quickSort(array: IntArray)
    = quickSort(array, 0, array.size - 1)

private fun quickSort(array: IntArray, start: Int, end: Int) {
    if (end > start) {
        val pivot = findGoodPivot(array, start, end)
        val newPivot = partition(array, start, end, pivot)
        quickSort(array, start, newPivot - 1)
        quickSort(array, newPivot + 1, end)
    }
}

private fun findGoodPivot(array: IntArray, start: Int, end: Int): Int {

    var aIdx = start
    var a = array[aIdx]
    var bIdx = (start + end) / 2
    var b = array[bIdx]
    var cIdx = end
    var c = array[cIdx]

    fun swapAB() {
        if (a > b) {
            a = b.also { b = a }
            aIdx = bIdx.also { bIdx = aIdx }
        }
    }
    swapAB()
    if (b > c) {
        b = c.also { c = b }
        bIdx = cIdx.also { cIdx = bIdx }
    }
    swapAB()

    return bIdx
}

private fun partition(array: IntArray, start: Int, end: Int, pivot: Int): Int {
    val pivotVal = array[pivot]
    swapInArray(array, pivot, end)
    var storeIdx = start
    for (i in start until end) {
        if (array[i] < pivotVal) { // array[i].CompareTo(pivotVal) < 0
            swapInArray(array, i, storeIdx)
            storeIdx++
        }
    }
    swapInArray(array, storeIdx, end)
    return storeIdx
}

private fun swapInArray(array: IntArray, i1: Int, i2: Int) {
    val temp = array[i1]
    array[i1] = array[i2]
    array[i2] = temp
}